### PR TITLE
feat/dpop

### PR DESCRIPTION
**Pull request by tusharpandey13**
- **feat:DPoP inital impl, tests are passing**
- **bugfix: add encryption valudation; misc edge case fixes, test updates**
- **chore:update docs**
- **chore:update docs in fetcher; add default import path for client**
- **feat:add dpop example**
- **feat:add clockskew and clockTolerance options]**
- **chore: fix example build**

### CODE REVIEW

**Comment by frederikprijck** [comment]
@claude

File: src/server/auth-client.ts [hunk 1]
```diff
@@ -1576,6 +1664,205 @@ export class AuthClient {
...
+      } else {
+        // For other BodyInit types, convert to string
+        requestBody = String(body);
+      }
+    }
+
+    // Make (DPoP)-authenticated request with retry logic for nonce errors
+    const protectedResourceRequestCall = () => {
+      return oauth.protectedResourceRequest(
+        session.tokenSet.accessToken,
```
Comment (frederikprijck) [inline src/server/auth-client.ts]: This will not be the correct token. How do we know which token the external URL needs?

Comment (tusharpandey13) [reply src/server/auth-client.ts]: This has been changed. Added `AccessTokenOptions` param to `fetchWithAuth()`, same is used to get an access token and that AT is used here.

Comment (frederikprijck) [reply src/server/auth-client.ts]: Why do we need `AccessTokenOptions ` ?

File: examples/with-dpop/api-server.js [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    console.log('[API] Step 1: Validating headers');
+    const authHeader = req.headers.authorization;
+    if (!authHeader) {
+      console.log('[API] Missing Authorization header');
+      return res.status(401).json({ error: 'Missing Authorization header' });
+    }
+
+    // Check if using DPoP scheme
+    if (!authHeader.toLowerCase().startsWith('dpop ')) {
+      console.log('[API] Invalid authorization scheme, expected DPoP but got:', authHeader.substring(0, 10));
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: examples/with-dpop/api-server.js [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    const dpopHeader = req.headers.dpop;
+    if (!dpopHeader) {
+      console.log('[API] Missing DPoP header');
+      return res.status(401).json({ error: 'Missing DPoP header' });
+    }
+
+    const accessToken = authHeader.slice(5); // Remove 'DPoP ' prefix
+    console.log('[API] Headers validated successfully, access token length:', accessToken.length);
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: examples/with-dpop/api-server.js [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+    // Check required DPoP proof claims
+    console.log('[API] Validating DPoP proof claims');
+    if (dpopProofPayload.htm !== req.method) {
+      console.log('[API] Method mismatch - expected:', req.method, 'got:', dpopProofPayload.htm);
+      return res.status(401).json({ error: 'DPoP proof htm mismatch' });
+    }
+
+    const expectedHtu = `${req.protocol}://${req.get('host')}${req.originalUrl.split('?')[0]}`;
+    console.log('[API] Expected HTU:', expectedHtu, 'Got:', dpopProofPayload.htu);
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: examples/with-dpop/api-server.js [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    return res.status(401).json({
+      error: 'DPoP validation failed',
+      message: error.message
+    });  
+  }
+};
+
+// Apply middleware based on DPoP configuration
+if (process.env.USE_DPOP === 'true') {
+  app.get('/api/shows', checkDpop, (req, res) => {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: examples/with-dpop/api-server.js [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    return next();
+  }
+
+  console.log('[API] DPoP enabled, starting validation process');
+
+  try {
+    // Step 1: Validate request has proper headers
+    console.log('[API] Step 1: Validating headers');
+    const authHeader = req.headers.authorization;
+    if (!authHeader) {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js:160]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: examples/with-dpop/api-server.js [hunk 7]
```diff
@@ -0,0 +1,316 @@
+require('dotenv').config({ path: './.env.local' });
+
+const express = require('express');
+const cors = require('cors');
+const morgan = require('morgan');
+const helmet = require('helmet');
+const { expressjwt: jwt } = require('express-jwt');
+const jwksRsa = require('jwks-rsa');
+const oauth = require('oauth4webapi');
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js:30]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: examples/with-dpop/api-server.js [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+const util = require('util');
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: examples/with-dpop/api-server.js [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        } else if (jwk.kty === 'RSA') {
+          alg = 'RS256';
+        } else {
+          console.error('[API] Unable to determine JWK algorithm for DPoP proof', { jwk });
+          return res.status(401).json({ error: 'Unable to determine JWK algorithm for DPoP proof' });
+        }
+      }
+      console.log('[API] Using algorithm:', alg);
+      const publicKey = await jose.importJWK(dpopProofHeader.jwk, alg);
+      const { payload: verifiedPayload } = await jose.jwtVerify(dpopHeader, publicKey, {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is a sample server file not shipped as part of nextjs-auth0 package.

File: src/server/auth-client.ts (hunk 1 above)
Comment (tusharpandey13) [reply src/server/auth-client.ts]: they supply scope, audience and if a refresh is needed. This solves the issue of DPoP always using the same token.

The user can now call `fetchWithAuth(info, {audience: "...", scope: "...", refresh: true});`

This should be `GetAccessTokenOptions` to include refresh

File: src/server/auth-client.ts [hunk 10]
```diff
@@ -380,6 +414,17 @@ export class AuthClient {
     authorizationParams.set("state", state);
     authorizationParams.set("nonce", nonce);
 
+    // Add dpop_jkt parameter if DPoP is enabled
+    if (this.dpopHandle && this.dpopKeyPair) {
+      try {
+        const publicKeyJwk = await jose.exportJWK(this.dpopKeyPair.publicKey);
+        const dpopJkt = await jose.calculateJwkThumbprint(publicKeyJwk);
+        authorizationParams.set("dpop_jkt", dpopJkt);
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/auth-client.ts:431]: We should throw a DPoPError here.

File: src/client/fetcher/fetcher.ts [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+    } as ProtectedRequestBody & { sessionData?: any };
+
+    const protectedRequestPath =
+      process.env.NEXT_PUBLIC_PROTECTED_REQUEST_ROUTE ||
+      "/auth/protected-request";
+
+    // Build the full URL for the protected request using baseUrl if available
+    const protectedRequestUrl = this.config.baseUrl
+      ? this.buildUrl(this.config.baseUrl, protectedRequestPath)
+      : protectedRequestPath;
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/client/fetcher/fetcher.ts]: this is outdated

File: src/server/client.ts [hunk 12]
```diff
@@ -857,6 +893,83 @@ export class Auth0Client {
...
+
+    // Override with init values if provided
+    if (init) {
+      if (init.method) method = init.method;
+      if (init.headers) headers = init.headers;
+      if (init.body !== undefined) body = init.body;
+    }
+
+    // Call the protected request logic directly (bypassing HTTP)
+    return this.authClient.executeProtectedRequest({
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/client.ts]: Deduplication implementation should be left to customer's logic since we are not making any intermediary http calls.

File: src/server/client.ts [hunk 13]
```diff
@@ -969,6 +1082,247 @@ export class Auth0Client {
...
+   */
+  private validateDpopConfiguration(options: Auth0ClientOptions): {
+    dpopKeyPair?: DpopKeyPair;
+    dpopOptions?: DpopOptions;
+  } {
+    const useDpop = options.useDpop || false;
+
+    // Build DPoP options with defaults from environment variables or provided options
+    const dpopOptions: DpopOptions = {
+      clockSkew:
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/client.ts:1180]: Added validation, docstring and warnings for the same. Update default values, moved to constants.

File: src/server/auth-client.ts [hunk 14]
```diff
@@ -235,6 +248,17 @@ export class AuthClient {
     // authorization server
     this.domain = options.domain;
     this.clientMetadata = { client_id: options.clientId };
+
+    // Apply DPoP timing validation options to client metadata
+    if (options.dpopOptions) {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/auth-client.ts]: Done

File: src/server/auth-client.ts [hunk 15]
```diff
@@ -1576,6 +1664,205 @@ export class AuthClient {
...
+        headers instanceof Headers ? headers : new Headers(headers);
+    }
+
+    // Convert body to oauth4webapi compatible type
+    let requestBody:
+      | import("oauth4webapi").ProtectedResourceRequestBody
+      | undefined;
+    if (body !== null && body !== undefined) {
+      // If it's already a ProtectedResourceRequestBody type, use it directly
+      if (
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/auth-client.ts]: We now directly use `ProtectedResourceRequestBody` in input contract of `fetchWithAuth`, byspassing the need for this whole logic.

File: src/server/auth-client.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/auth-client.ts:431]: Done, error classes have been added.

File: src/server/client.ts [hunk 16]
```diff
@@ -969,6 +1082,247 @@ export class Auth0Client {
...
+   * by attempting to sign and verify a test message.
+   * @returns true if keys are compatible, false otherwise
+   */
+  private validateKeyPairCompatibility(
+    privateKey: CryptoKey | any,
+    publicKey: CryptoKey | any
+  ): boolean {
+    try {
+      // Create test data
+      const testData = "test-data-for-key-pair-validation";
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/client.ts:1299]: Done.

File: src/server/client.ts [hunk 17]
```diff
@@ -969,6 +1082,247 @@ export class Auth0Client {
...
+
+    // If DPoP is enabled but no keypair provided, check if environment variables exist
+    if (useDpop) {
+      const privateKeyPem = process.env.AUTH0_DPOP_PRIVATE_KEY;
+      const publicKeyPem = process.env.AUTH0_DPOP_PUBLIC_KEY;
+
+      if (privateKeyPem && publicKeyPem) {
+        try {
+          // Load and convert keys synchronously
+          const privateKeyNodeJS = createPrivateKey(privateKeyPem);
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/client.ts:1222]: Added docstring explaining the same.

File: src/server/auth-client.ts [hunk 18]
```diff
@@ -288,6 +312,11 @@ export class AuthClient {
     this.enableAccessTokenEndpoint = options.enableAccessTokenEndpoint ?? true;
     this.noContentProfileResponseWhenUnauthenticated =
       options.noContentProfileResponseWhenUnauthenticated ?? false;
+
+    if (options.dpopKeyPair && (options.useDpop ?? false)) {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/auth-client.ts]: done.

File: src/server/auth-client.ts [hunk 19]
```diff
@@ -1576,6 +1664,205 @@ export class AuthClient {
...
+        }
+        requestBody = params;
+      } else {
+        // For other BodyInit types, convert to string
+        requestBody = String(body);
+      }
+    }
+
+    // Make (DPoP)-authenticated request with retry logic for nonce errors
+    const protectedResourceRequestCall = () => {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply src/server/auth-client.ts]: done.

File: examples/with-dpop/api-server.js [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+    return next();
+  }
+
+  console.log('[API] DPoP enabled, starting validation process');
+
+  try {
+    // Step 1: Validate request has proper headers
+    console.log('[API] Step 1: Validating headers');
+    const authHeader = req.headers.authorization;
+    if (!authHeader) {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js:160]: This is no longer relevant.
Additionaly, this file is part of an example app, not prod code.

File: examples/with-dpop/api-server.js [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+    }
+
+    // Check if using DPoP scheme
+    if (!authHeader.toLowerCase().startsWith('dpop ')) {
+      console.log('[API] Invalid authorization scheme, expected DPoP but got:', authHeader.substring(0, 10));
+      return res.status(401).json({ error: 'Expected DPoP authorization scheme' });
+    }
+
+    const dpopHeader = req.headers.dpop;
+    if (!dpopHeader) {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js:172]: This is no longer relevant.
Additionaly, this file is part of an example app, not prod code.

File: examples/with-dpop/api-server.js [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+      console.log('[API] Missing DPoP header');
+      return res.status(401).json({ error: 'Missing DPoP header' });
+    }
+
+    const accessToken = authHeader.slice(5); // Remove 'DPoP ' prefix
+    console.log('[API] Headers validated successfully, access token length:', accessToken.length);
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is no longer relevant.
Additionaly, this file is part of an example app, not prod code.

File: examples/with-dpop/api-server.js [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+
+    // Check required DPoP proof claims
+    console.log('[API] Validating DPoP proof claims');
+    if (dpopProofPayload.htm !== req.method) {
+      console.log('[API] Method mismatch - expected:', req.method, 'got:', dpopProofPayload.htm);
+      return res.status(401).json({ error: 'DPoP proof htm mismatch' });
+    }
+
+    const expectedHtu = `${req.protocol}://${req.get('host')}${req.originalUrl.split('?')[0]}`;
+    console.log('[API] Expected HTU:', expectedHtu, 'Got:', dpopProofPayload.htu);
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is no longer relevant.
Additionaly, this file is part of an example app, not prod code.

File: examples/with-dpop/api-server.js [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+    return res.status(401).json({
+      error: 'DPoP validation failed',
+      message: error.message
+    });  
+  }
+};
+
+// Apply middleware based on DPoP configuration
+if (process.env.USE_DPOP === 'true') {
+  app.get('/api/shows', checkDpop, (req, res) => {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is no longer relevant.
Additionaly, this file is part of an example app, not prod code.

File: examples/with-dpop/api-server.js [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
+        } else if (jwk.kty === 'RSA') {
+          alg = 'RS256';
+        } else {
+          console.error('[API] Unable to determine JWK algorithm for DPoP proof', { jwk });
+          return res.status(401).json({ error: 'Unable to determine JWK algorithm for DPoP proof' });
+        }
+      }
+      console.log('[API] Using algorithm:', alg);
+      const publicKey = await jose.importJWK(dpopProofHeader.jwk, alg);
+      const { payload: verifiedPayload } = await jose.jwtVerify(dpopHeader, publicKey, {
```
[the remark this answers is not in the document]
Comment (tusharpandey13) [reply examples/with-dpop/api-server.js]: This is no longer relevant.
Additionaly, this file is part of an example app, not prod code.

**Comment by tusharpandey13** [comment]
@claude

**Comment by tusharpandey13** [comment]
@claude

**Comment by tusharpandey13** [comment]
Superceded by #2357
